gentypes.go

```diff
@@ -0,0 +1,211 @@
+		}
+		str := path[0]
+		if !strings.HasPrefix(str, "map of") {
+			return str, nil
+		}
+		str = strings.TrimPrefix(str, "map of")
+		str = strings.Trim(str, " ")
+		m := strings.Split(str, ":")
+		if len(m) != 2 {
+			shouldStr := `should be map of KeyType:ValueType`
+			return "", fmt.Errorf("bad map definition '%s' (%s)", str, shouldStr)
+		}
+		return fmt.Sprintf("map[%s]%s", m[0], m[1]), nil
+	}
+	return "", fmt.Errorf("type %v is neither primitive nor reference", t)
+
+}
+
+// GetLine returns the line for a given Sysl type from its SourceContext
+func GetLine(t *pb.Type) (int32, error) {
+	if t.GetPrimitive() != pb.Type_NO_Primitive || t.GetTypeRef() != nil {
+		return t.GetSourceContext().Start.Line, nil
+	} else if t.GetList() != nil {
+		return t.GetList().GetType().GetSourceContext().Start.Line, nil
+	} else if t.GetSet() != nil {
+		return t.GetSet().GetSourceContext().Start.Line, nil
+	} else if t.GetTuple() != nil {
+		for _, t2 := range t.GetTuple().GetAttrDefs() {
+			return GetLine(t2)
+		}
+	}
+	return 0, fmt.Errorf("unknown type %v for getting line", t)
+}
+
+//NamesSortedBySourceContext sorts the keys of the input types according to occurrence
+// in Sysl definition file, derived from SourceContext
+func NamesSortedBySourceContext(types map[string]*pb.Type) ([]string, error) {
+	type lineName struct {
+		name string
+		line int32
+	}
+	size := len(types)
+	lineNames := make([]lineName, size)
+	i := 0
+	for name, t := range types {
+		line, err := GetLine(t)
+		if err != nil {
+			return nil, err
+		}
+		lineNames[i] = lineName{name, line}
+		i++
+	}
+	sort.Slice(lineNames, func(i, j int) bool {
+		return lineNames[i].line < lineNames[j].line
+	})
+	result := make([]string, size)
+	for i := 0; i < size; i++ {
+		result[i] = lineNames[i].name
+	}
+	return result, nil
+}
+
+// SplitUppercase splits into fields at all upper case letters and converts
+// fields to lower case
+func SplitUppercase(str string) []string {
+	result := []string{}
+	var i int
+	for s := str; s != ""; s = s[i:] {
+		i = strings.IndexFunc(s[1:], unicode.IsUpper) + 1
+		if i == 0 {
+			i = len(s)
+		}
+		result = append(result, strings.ToLower(s[:i]))
+	}
+	return result
+}
+
+// GetJSONProperty extracts JSON property name from attribute and defaults to type name
+func GetJSONProperty(name string, t *pb.Type, sep string) string {
+	if attrOverride, ok := t.Attrs["json"]; ok {
+		return attrOverride.GetS()
+	}
+	if sep == "" {
+		return name
+	}
+	return strings.Join(SplitUppercase(name), sep)
+}
+
+// GenStructField creates a single line inside a struct definition
+func GenStructField(fieldName string, fieldType *pb.Type, sep string) (string, error) {
+	var typeStr string
+	var err error
+	if fieldType.GetPrimitive() != pb.Type_NO_Primitive || fieldType.GetTypeRef() != nil {
+		// Primitive type, reference or map
+		if typeStr, err = GenSimpleType(fieldType); err == nil {
+			typeStr = fmt.Sprintf("%s %s ", fieldName, typeStr)
+		}
```

Comment:
Could end up with empty typeStr? Ditto below

Reply:
I'm not quite sure I understand. I've added a catch all at the end:
```go
if fieldStr == "" {
		return "", fmt.Errorf("unknown type %s", fieldType.String())
}
```
In my coming up PR this is factored out to it's own func...